tests/test-dladdr.c

```diff
@@ -0,0 +1,185 @@
+    int result = dladdr( addr, &info );
+    int passed = 0;
+    if (!result)
+    {
+        passed = expected_result == NoInfo || expected_result == Fail;
+    }
+    else
+    {
+        int sym_match  = info.dli_sname && strcmp( addrsym, info.dli_sname ) == 0;
+        int addr_match  = addr == info.dli_saddr;
+        passed = (expected_result == Pass && sym_match && addr_match)
+                 || (expected_result == PassWithoutSymbol && addr_match && !info.dli_sname)
+                 || (expected_result == PassWithDifferentAddress && sym_match && !addr_match)
+                 || (expected_result == Fail && !(sym_match && addr_match));
+    }
+    printf( "check address %p which has symbol '%s' -> %s%s", addr, addrsym, passed ? "passed" : "failed", verbose || !passed ? " " : "\n" );
+    if( verbose || !passed )
+    {
+        if( !result )
+            printf( "(could not get symbol information for address %p)\n", addr );
+        else
+            print_dl_info( &info, "(", ")\n");
+    }
+    return !passed;
+}
+
+/**
+ * @brief return address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @return pointer to symbol address or NULL
+ */
+void *get_symbol_address( char *libname, char *sym )
+{
+    void *library = NULL;
+    void *addr = NULL;
+
+    library = dlopen( libname, RTLD_GLOBAL );
+    if ( library == NULL )
+    {
+        fprintf( stderr, "could not open '%s'\n", libname );
+        return NULL;
+    }
+
+    addr = dlsym( library, sym );
+    dlclose( library );
+    if (!addr)
+        fprintf( stderr, "could not get address for symbol '%s'\n", sym );
+    return addr;
+}
+
+/**
+ * @brief check address from a symbol located in a shared lilbrary
+ * @param libname librray to get the address from
+ * @param addrsym symbol to get the address for
+ * @param should_match result should match the given values
+ * @return 0 check passed
+ * @return 1 check failed
+ */
+int check_dladdr_by_dlopen( char *libname, char *sym, int should_match )
+{
+    void *addr = get_symbol_address( libname, sym);
+    if (!addr)
+        return 1;
+    return check_dladdr( addr, sym, should_match );
+}
+
+#ifdef _WIN32
+#include <windows.h>
+#if defined(__GNUC__) && (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6))
+/* hide warning "reclared without 'dllimport' attribute" */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wattributes"
+#endif
+/*
+ * Windows API functions decorated with WINBASEAPI are imported from iat.
+ * For testing force linking by import thunk for the following functions
+ */
+HMODULE WINAPI GetModuleHandleA (LPCSTR lpModuleName);
+SIZE_T WINAPI VirtualQuery (LPCVOID lpAddress, PMEMORY_BASIC_INFORMATION lpBuffer, SIZE_T dwLength);
+/* force linking by iat */
+WINBASEAPI HMODULE WINAPI LoadLibraryExA(LPCSTR lpLibFileName, HANDLE hFile, DWORD  dwFlags);
+WINBASEAPI int __cdecl atoi(const char *_Str);
```

Comment:
`WINBASEAPI` does not work for `i586-mingw32msvc-gcc` compiler.

With `WINBASEAPI HMODULE WINAPI LoadLibraryExA(LPCSTR lpLibFileName, HANDLE hFile, DWORD  dwFlags);` test is failing:
```
check address 00402270 which has symbol 'LoadLibraryExA' -> failed (filename: C:\windows\system32\KERNEL32.dll base: 7B420000 symbol name: 'LoadLibraryExA' addr: 7B45CB80)
```

And with `__declspec(dllimport) HMODULE WINAPI LoadLibraryExA(LPCSTR lpLibFileName, HANDLE hFile, DWORD  dwFlags);` test is passing:
```
check address 7B45CB80 which has symbol 'LoadLibraryExA' -> passed
```

So please change `WINBASEAPI` to `__declspec(dllimport)`.

Reply:
I created pull request https://github.com/dlfcn-win32/dlfcn-win32/pull/88 to run Travis tests also under i586-mingw32msvc-gcc compiler to catch such problems...

Reply:
And seems that first 8 failed tests on https://ci.appveyor.com/project/dlfcn-win32/dlfcn-win32/builds/36790613 would be similar issue that WINBASEAPI does not work... It looks like that i586-mingw32msvc-gcc does not use nor define WINBASEAPI, it just empty.

Reply:
WINBASEAPI is the official name for that macro - I will add a fallback if it is not present.

Reply:
Seems that this fallback does not work, see: https://travis-ci.com/github/dlfcn-win32/dlfcn-win32/jobs/459175512

Can you really change code to `__declspec(dllimport) HMODULE WINAPI LoadLibraryExA(LPCSTR lpLibFileName, HANDLE hFile, DWORD dwFlags);`?

Reply:
That macro is defined by the ming header packages, On opensuse there is
``` 
rpm -q  mingw32-headers  --info
Name        : mingw32-headers
Version     : 7.0.0
Release     : lp151.5.1
...
Install Date: Mo 16 Nov 2020 13:40:42 CET
Group       : Development/Libraries/C and C++
Source RPM  : mingw32-headers-7.0.0-lp151.5.1.src.rpm
Build Date  : Mo 04 Mai 2020 21:55:01 CEST
Vendor      : obs://build.opensuse.org/windows:mingw
URL         : http://mingw-w64.sf.net/
Summary     : MinGW-w64 headers for Win32 and Win64
Description :
MinGW-w64 delivers runtime, headers and libs for developing both 64
bit (x64) and 32 bit (x86) windows applications using GCC and other
free software compilers.

This subpackage contains the header files.
Distribution: windows:mingw:win32 / openSUSE_Leap_15.1
```
Can you tell me what version of headers are used by that compiler and where they come from ?

Reply:
Now it started passing! https://ci.appveyor.com/project/dlfcn-win32/dlfcn-win32/builds/36819915/job/bsvacwhccjek93vr

I was looking at it and problem is that these Mingw versions defines WINBASEAPI macro, but based on other (internal) macros. And by default they do not use dllimport and do linking via import thunks. WINBASEAPI defines how system base functions are linked. And NOT it enforce usage of dllimport.